### PR TITLE
docs: Add comments grid iterator

### PR DESCRIPTION
**Pull request by CarloVarni**
This adds some comments to the code to describe what the grid iterator classes do. It also adds two methods for retrieving the current local and global positions from the iterators, as well as adding some unit tests to check these newly-added methods.

### CODE REVIEW

File: Core/include/Acts/Utilities/GridIterator.ipp
Comment (paulgessinger) [inline Core/include/Acts/Utilities/GridIterator.ipp:1]: Doxygen style triple-slash comments don't to anything special inside functions. Consider making this a regular double-slash code comment (I think it's worth having these comments!)

Comment (CarloVarni) [reply Core/include/Acts/Utilities/GridIterator.ipp:1]: done

File: Core/include/Acts/Utilities/GridIterator.hpp [hunk 1]
```diff
@@ -27,46 +31,135 @@ class GridGlobalIterator {
...
+  /// @brief Create incremented iterator
+  /// @param [in] offset The increment value
+  /// @return The incremented iterator
   GridGlobalIterator<T, Axes...> operator+(const std::size_t offset) const;
+  /// @brief Create decremented iterator
+  /// @param [in] offset The decrement value
+  /// @return The decremented iterator
   GridGlobalIterator<T, Axes...> operator-(const std::size_t offset) const;
 
+  /// @brief Distance between two GridGlobalIterator
```
Comment (niermann999) [inline Core/include/Acts/Utilities/GridIterator.hpp:115]:
```suggestion
  /// @brief Distance between two GridGlobalIterators
```

File: Core/include/Acts/Utilities/GridIterator.hpp [hunk 2]
```diff
@@ -78,47 +171,131 @@ class GridLocalIterator {
...
+  /// @pre None of the navigation vectors is allowed to be an empty vector
   GridLocalIterator(Acts::Grid<T, Axes...>&& grid,
                     const std::array<std::size_t, DIM>& indexes,
                     std::array<std::vector<std::size_t>, DIM> navigation) =
       delete;
+  /// @brief Constructor
+  /// @param [in] grid The grid
+  /// @param [in] indexes The local position
+  ///
+  /// @pre The local bins representing the local bins must be a valid local position in the grid
```
Comment (niermann999) [inline Core/include/Acts/Utilities/GridIterator.hpp:195]:
```suggestion
  /// @pre The local bins must be a valid local position in the grid
```

File: Core/include/Acts/Utilities/GridIterator.hpp [hunk 3]
```diff
@@ -78,47 +171,131 @@ class GridLocalIterator {
...
   GridLocalIterator(const Acts::Grid<T, Axes...>& grid,
                     const std::array<std::size_t, DIM>& indexes);
+  /// @brief Constructor with custom navigation pattern
+  /// @param [in] grid The grid
+  /// @param [in] indexes The local position
+  /// @param [in] navigation The custom navigation pattern for each axis
+  ///
+  /// @pre The local bins representing the local bins must be a valid local position in the grid.
```
Comment (niermann999) [inline Core/include/Acts/Utilities/GridIterator.hpp:203]:
```suggestion
  /// @pre The local bins must be a valid local position in the grid.
```

File: Core/include/Acts/Utilities/GridIterator.hpp [hunk 4]
```diff
@@ -78,47 +171,131 @@ class GridLocalIterator {
...
+  /// This represent the position in the navigation pattern.
+  /// For each axis, the current index goes from 0ul to the size of the
+  /// corresponding navigation
+  ///
+  /// The local position in the gris is then obtained, for each axis i,
+  /// via m_navigationIndex[m_currentIndex[i]]
   std::array<std::size_t, DIM> m_currentIndex{};
+  /// @brief The custom navigation pattern in the grid
+  ///
+  /// This allows user to define any custom iteration sequence in all the
```
Comment (niermann999) [inline Core/include/Acts/Utilities/GridIterator.hpp:295]:
```suggestion
  /// This allows users to define any custom iteration sequence in all the
```

File: Core/include/Acts/Utilities/GridIterator.hpp [hunk 5]
```diff
@@ -27,46 +31,135 @@ class GridGlobalIterator {
...
   GridGlobalIterator<T, Axes...>& operator++();
+  /// @brief Increment operator (post)
+  /// @return The global iterator before the increment
+  ///
+  /// This will increase the global position by one
   GridGlobalIterator<T, Axes...> operator++(int);
 
+  /// @brief Retrieve the global position
+  /// @return The current global position in the grid
+  std::size_t globalPosition() const;
```
Comment (niermann999) [inline Core/include/Acts/Utilities/GridIterator.hpp:141]: Tbh, ```globalPosition``` sounds a bit like a position in space. Maybe ```globalBinIndex``` would be clearer?

**Review by niermann999 — COMMENTED** [review]
Looks good in general

**Comment by CarloVarni** [comment]
Hi @niermann999 thanks for the comment. I have applied your suggestions

**Review by niermann999 — APPROVED** [review]

**Comment by CarloVarni** [comment]
Hi @niermann999 I had to make a minor change to fix a problem in the doc test (`@brief` -> `@param`)

This PR needs another approval for the CI to be happy

**Review by niermann999 — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [5893004534ae1bcae4ac7fab56c723710c67db25]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164049)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164048)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164047)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164046)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164045)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164044)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164043)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164042)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164041)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164040)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164039)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164038)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsGSFRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164037)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164036)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164035)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164034)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164033)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164032)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164031)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164030)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164029)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164028)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35164027)||

**Pull request merged**